torch/_dynamo/optimizations/analysis.py

```diff
@@ -122,9 +122,27 @@ def has_mutation(gm, example_inputs, inputs_only=False):
     # TODO - moco gives bad accuracy with Aliasing. gm is getting mutated in a bad way.
 
     if fake_tensors_available and config.fake_tensor_propagation:
-        with FakeTensorMode() as fake_mode:
+
+        def _wrap_to_fake_tensor(t, *, f_mode):
+            if type(t) in (torch.Tensor, torch.nn.Parameter):
+                static_shapes_ = config.dynamic_shapes is False
+                return fake_mode.from_tensor(
+                    t, static_shapes=config.dynamic_shapes is not False
+                )
+            else:
+                return t
```

Comment:
You're partially evaluating this right at the definition site, why not just close over `fake_mode`

Reply:
Just to confirm I'm understanding this comment - the argument is that further down we just created the `FakeTensorMode()` so don't need to bother with `functools.partial(_wrap_to_fake_tensor, f_mode=fake_mode)`, and can just directly use `fake_mode` from the outer context? If that's what you're saying then sounds fine to me.

As for "why are we creating a fresh FakeTensorMode" (maybe this wasn't the q though) - this is the annoying thing that is described in the comments on line 135. The analysis pass should properly give all of the fake tensor inputs dynamic shapes, because the graph can have SymInt-related ops in it like `x.__floordiv__(y)` that normal ints will break on. But we don't want to use the existing, global `FakeTensorMode` that dynamo has, because that might install extra guards. And we don't care about installing shape guards during this analysis pass (which would be redundant anyway).

Reply:
closing